### PR TITLE
expression: rewrite builtin string function: REVERSE

### PR DESCRIPTION
**Pull request by zz-jason**
fix https://github.com/pingcap/tidb/issues/4043
to https://github.com/pingcap/tidb/issues/4080

### CODE REVIEW

File: expression/builtin_string.go [hunk 1]
```diff
@@ -490,33 +490,31 @@ type reverseFunctionClass struct {
...
-	return sig.setSelf(sig), errors.Trace(c.verifyArgs(args))
+	if err := c.verifyArgs(args); err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	bf, err := newBaseBuiltinFuncWithTp(args, ctx, tpString, tpString)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	*bf.tp = *args[0].GetType()
```
Comment (shenli) [inline expression/builtin_string.go]: How about argument is int?
mysql> select reverse(123);
+--------------+
| reverse(123) |
+--------------+
| 321          |
+--------------+
1 row in set (0.00 sec)

mysql>

Comment (zz-jason) [inline expression/builtin_string.go]: we will cast 123(int) to "123" and reverse it to "321". performs the same as mysql

**Comment by tiancaiamao** [comment]
CI fail.

**Comment by XuHuaiyu** [comment]
please fix the ci @zz-jason

File: expression/builtin_string.go [hunk 2]
```diff
@@ -490,33 +490,57 @@ type reverseFunctionClass struct {
...
+	if types.IsBinaryStr(bf.tp) {
+		sig := &builtinReverseBinarySig{baseStringBuiltinFunc{bf}}
+		return sig.setSelf(sig), nil
```
Comment (XuHuaiyu) [inline expression/builtin_string.go]: remove this line, use `else`, and use `return` at the outside.

File: expression/builtin_string.go [hunk 3]
```diff
@@ -490,33 +490,57 @@ type reverseFunctionClass struct {
...
+
+// evalString evals a REVERSE(str).
+// See https://dev.mysql.com/doc/refman/5.7/en/string-functions.html#function_reverse
+func (b *builtinReverseBinarySig) evalString(row []types.Datum) (string, bool, error) {
+	str, isNull, err := b.args[0].EvalString(row, b.ctx.GetSessionVars().StmtCtx)
+	if isNull || err != nil {
+		return "", true, errors.Trace(err)
+	}
+	strBytes := []byte(str)
+	for i, length := 0, len(strBytes); i < length/2; i++ {
```
Comment (XuHuaiyu) [inline expression/builtin_string.go]: this seems duplicate with `char()` https://github.com/pingcap/tidb/pull/4041/files#diff-314e997a9df9b116e8f0aad4149df468R1451

File: expression/integration_test.go [hunk 4]
```diff
@@ -86,6 +86,22 @@ func (s *testIntegrationSuite) TestFuncREPEAT(c *C) {
 	r.Check(testkit.Rows("<nil> <nil> <nil> <nil> <nil> <nil>"))
 }
 
+func (s *testIntegrationSuite) TestFuncReverse(c *C) {
```
Comment (XuHuaiyu) [inline expression/integration_test.go]: move this to TestStringBuiltin

File: util/stringutil/string_util.go [hunk 5]
```diff
@@ -24,15 +24,6 @@ import (
 // ErrSyntax indicates that a value does not have the right syntax for the target type.
 var ErrSyntax = errors.New("invalid syntax")
 
-// Reverse returns its argument string reversed rune-wise left to right.
```
Comment (XuHuaiyu) [inline util/stringutil/string_util.go]: we can maintain `reverseBinaryStr` and `reverseNonBinaryStr` here?

**Comment by zz-jason** [comment]
"continuous-integration/travis-ci/pr" failed, investigating, WIP

**Comment by zz-jason** [comment]
@XuHuaiyu @tiancaiamao @breeswish @winkyao PTAL

File: expression/builtin_string.go (hunk 1 above)
Comment (tiancaiamao) [inline expression/builtin_string.go]: why `*bf.tp = *args[0].GetType()` not `bf.tp = args[0].GetType()`

**Comment by tiancaiamao** [comment]
LGTM

File: expression/builtin_string.go (hunk 1 above)
Comment (zz-jason) [inline expression/builtin_string.go]: it's better to have their own `FieldType` for every `builtinFunc`

**Pull request merged**